tripy/docs/pre0_user_guides/02-compiler.md

```diff
@@ -0,0 +1,169 @@
+# Using the Compiler
+
+```{contents} Table of Contents
+:depth: 3
+```
+
+## Walk through model compilation and deployment
+
+Let's walk through a simple example of a [GEGLU](https://arxiv.org/abs/2002.05202v1) module defined below:
+
+```py
+# doc: no-eval
+class GEGLU(tp.Module):
+    def __init__(self, dim_in, dim_out):
+        self.proj = tp.Linear(dim_in, dim_out * 2)
+        self.dim_out = dim_out
+    
+    def __call__(self, x):
+        proj = self.proj(x)
+        x, gate = tp.split(proj, 2, proj.rank - 1)
+        return x * tp.gelu(gate)
+```
+
+To naively run `GEGLU` in eager mode:
+
+```py
+# doc: no-eval
+layer = GEGLU(2, 8)
+inp = tp.ones((1, 2))
+out = layer(inp)
+print(out)
+```
+
+Now, let's try to optimize this model for inference using Tripy's Compiler.
+
+First, let's initialize the compiler with `layer` which lets the `Compiler` know the properties of the function we are compiling (e.g. function signature).
+
+```py
+# doc: no-eval
+compiler = tp.Compiler(layer)
+```
+Next, we need to provide information about each input using `tp.InputInfo`. The first argument for `InputInfo` is `shape`, where we specify either the static or dynamic shape information for each dimension. In the example below, we assume the shape of `inp` is static (`(1, 2)`). The second argument specifies the `dtype` for the input:
+
+```py
+# doc: no-eval
+inp_info = tp.InputInfo(shape=(1, 2), dtype=tp.float32)
+```
+Now, we can call the `compile` function to obtain a compiled function and use it for inference:
+
+```py
+# doc: no-eval
+fast_geglu = compiler.compile(inp_info)
+print(fast_geglu(inp))
+```
+
+### Optimization profiles
+
+In the example above, we assumed `inp` has a static shape of `(1, 2)`. Now, let’s assume that the shape of `inp` can vary from `(1, 2)` to `(16, 2)`. To express this constraint to the compiler, we can provide the range of shapes to `InputInfo` using `shape=((1, 8, 16), 2)`. This indicates to the compiler that the first dimension can vary from 1 to 16, and it should optimize for performance based on 8.
+
+```py
+# doc: no-eval
+inp_info = tp.InputInfo(shape=((1, 8, 16), 2), dtype=tp.float32)
+fast_geglu = compiler.compile(inp_info)
+print(fast_geglu(inp))
+
+# Let's change the shape of input to (2, 2)
+inp = tp.Tensor([[1., 2.], [2., 3.]], dtype=tp.float32)
+print(fast_geglu(inp))
+```
+
+### Errors for out of bounds inference
+If the user provides input that does not comply with the dynamic shape constraint given to the compiler, `Tripy` will produce an error with relevant information:
+
+```py
+# doc: no-eval
+from tests import helper
```

Comment:
Can we avoid the internal helper? We can add a `# doc: allow-exception` tag or something like that instead. That way the text of the exception can also be appended automatically